### PR TITLE
[Console] Allow Application to handle PHP 7 Errors

### PR DESCRIPTION
**Pull request by ogizanagi**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | not yet

This allows to handle nicely any `\Throwable` when using the console component standalone with PHP 7.

With the full-stack framework, a throwable is caught and handled by the Debug component `ErrorHandler` and the exception handler registered by `Symfony\Component\HttpKernel\EventListener\DebugHandlersListener`. (see https://github.com/symfony/symfony/pull/20797)

### CODE REVIEW

File: src/Symfony/Component/Console/Application.php [hunk 1]
```diff
@@ -272,6 +279,22 @@ public function setCatchExceptions($boolean)
     }
 
     /**
+     * @return bool Whether any throwable is caught or not during commands execution
+     */
+    public function areThrowablesCaught()
+    {
+        return $this->catchExceptions;
```
Comment (chalasr) [inline src/Symfony/Component/Console/Application.php]: should be `return $this->catchThrowables` I guess

File: src/Symfony/Component/Console/Tests/ApplicationTest.php [hunk 2]
```diff
@@ -1000,6 +1000,31 @@ public function testRunWithErrorFailingStatusCode()
...
+    /**
+     * @requires PHP 7
+     */
+    public function testRunWithErrorCatchThrowables()
+    {
+        $application = new Application();
+        $application->setCatchThrowables(true);
+        $application->setAutoExit(false);
+
+        $this->assertTrue($application->areThrowablesCaught());
```
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: maybe a test should be added for ensuring they are not caught by default?

**Comment by chalasr** [comment]
It's surely debatable but I would use `$catchErrors` instead of `$catchThrowables` since exceptions are throwable.

File: src/Symfony/Component/Console/Application.php [hunk 3]
```diff
@@ -615,10 +638,10 @@ public static function getAbbreviations($names)
     /**
      * Renders a caught exception.
      *
-     * @param \Exception      $e      An exception instance
-     * @param OutputInterface $output An OutputInterface instance
+     * @param \Throwable|\Exception $e      An throwable instance
+     * @param OutputInterface       $output An OutputInterface instance
      */
-    public function renderException(\Exception $e, OutputInterface $output)
+    public function renderException(/*\Throwable */$e, OutputInterface $output)
```
Comment (stof) [inline src/Symfony/Component/Console/Application.php]: removing the typehint is a BC break for classes extending the method (and there are such classes out there)

File: src/Symfony/Component/Console/Application.php [hunk 4]
```diff
@@ -62,6 +62,7 @@ class Application
     private $name;
     private $version;
     private $catchExceptions = true;
+    private $catchThrowables = false;
```
Comment (stof) [inline src/Symfony/Component/Console/Application.php]: I don't think having a second flag makes sense.

Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: What will you suggest instead?
We still have to differentiate exceptions from other throwables, as the last ones are currently handled in the fullstack by an exception handler, whereas exceptions are handled by the Console component itself.

We could make `catchExceptions` accept bit flags (`Application::CATCH_EXCEPTIONS`, `Application::CATCH_THROWABLES`) and consider `true` the same as `Application::CATCH_EXCEPTIONS` only. But I don't like it semantically and would complicate things.

File: src/Symfony/Component/Console/Application.php (hunk 3 above)
Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: Indeed. For now I've deprecated the `renderException` method in favor of `renderThrowable`.

The typehint is still commented and would be uncommented as soon as the PHP 5 support is dropped. I guess it's likely to happen for 4.0, so I can add it to `UPGRADE-4.0.md` file ?

**Comment by ogizanagi** [comment]
> It's surely debatable but I would use $catchErrors instead of `$catchThrowables` since exceptions are throwable.

I asked myself about that. But that might be misleading for PHP 5 users (actual php errors won't be handled). Also throwables does not mean a PHP error, but any other class implementing `\Throwable`.

**Comment by chalasr** [comment]
> Also throwables does not mean a PHP error, but any other class implementing \Throwable.

From the PHP manual:

> PHP classes cannot implement the Throwable interface directly

Which class being neither a subclass of `\Error` nor `\Exception` could be throwable?

**Comment by ogizanagi** [comment]
> PHP classes cannot implement the Throwable interface directly

You're right. Forget about this argument.

Still it might be misleading for PHP 5 users. Unless we register `renderThrowable/Error` as an error handler if `catchThrowables/catchErrors` is `true` ?

**Comment by ogizanagi** [comment]
How bad would it be to handle errors by using the `Debug` component as in 310eaa1f28dad62310126cb10b7bcd2437093ece ? Calling `Debug::enable` outside from a bootstrapping file looks hazardous to me, but what could be proposed instead, or how to make it safer/reliable? Should we consider that the end-user opted-in by using running this application with `catchErrors` and accept it may have a wider effect?
We should probably not do anything for PHP 5 after all...

**Comment by fabpot** [comment]
I would not rename `renderException()`. The whole `Throwable`/`Error`/`Exception` hierarchy is there in PHP for BC. There is no need to expose that to our users. At the end of the day, they are dealing with exceptions.

**Comment by ogizanagi** [comment]
@fabpot : Fair enough. Last chance for this then is to create a special exception wrapping an `\Error`, allowing to pass it to `renderException`.
Eventually, the typehint will change in 4.0 to use `\Throwable`.

**Comment by ogizanagi** [comment]
There is no more BC break nor deprecation here AFAIK. So the label can be removed :)

The typehint will however change in 4.0 without BC layer between, as it was done in 3.0 (https://github.com/symfony/symfony/pull/13756). Or we can keep the internal `ErrorException` and keep the typehint on `\Exception`.

File: src/Symfony/Component/Console/Application.php [hunk 5]
```diff
@@ -272,6 +281,22 @@ public function setCatchExceptions($boolean)
...
+        return $this->catchErrors;
+    }
+
+    /**
+     * @param bool $boolean Whether to catch errors or not during commands execution
+     */
+    public function setCatchErrors($boolean)
+    {
+        $this->catchErrors = (bool) $boolean;
+    }
```
Comment (fabpot) [inline src/Symfony/Component/Console/Application.php]: The addition of these 2 methods does not seem related to this PR and should be removed.

Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: I doubt about the usefulness of `areErrorsCaught` right now, but this echoes `areExceptionsCaught`.
`setCatchErrors` is the way to tell the application instance to catch php 7 errors and cannot be removed.

IMHO, we still have to differentiate exceptions from errors, as the last ones are currently handled in the fullstack framework by the handler registered through the Debug component and the `DebugHandlersListener` from HttpKernel, whereas classic exceptions are handled by the Console component itself. Thus, relying on `catchExceptions` only will change the behavior using the full stack, which can be annoying and a BC break. The handler used in full stack does extra work on the exception. For instance:

Using full stack:

>  [Symfony\Component\Debug\Exception\UndefinedMethodException]                                                                  
  Attempted to call an undefined method named "Option" of class "Symfony\Component\Console\Input\ArgvInput".                    
  Did you mean to call e.g. "getOption", "getOptions", "getParameterOption", "hasOption", "hasParameterOption" or "setOption"?

versus the console component standalone, with the `catchErrors` flag:

> [Error]                                                                       
  Call to undefined method Symfony\Component\Console\Input\ArgvInput::Option()

**Comment by ogizanagi** [comment]
I've just realized the debug component was already required by the console component, thus I don't need the internal `Symfony\Component\Console\Exception\ErrorException` but can directly use `FatalThrowableError` (apparently `symfony/debug` is only required for this class).

I've also removed the notice in `UPGRADE-4.0.md`, unless we indeed decide to handle `\Throwable` in `Application::renderException()`.

**Comment by chalasr** [comment]
Rethinking about this, I think the `$catchErrors` flag should be reverted and errors should always be converted and handled as any exception.

**Comment by ogizanagi** [comment]
Still doesn't look sensible to me to loose informations brung by the `Symfony\Component\Debug\ErrorHandler` and output by the console thanks to the `Symfony\Component\HttpKernel\EventListener\DebugHandlersListener` (as mentioned in https://github.com/symfony/symfony/pull/20808#discussion_r91832420).

If I apply the following patch:

<details>
<summary><strong>PATCH 1 (handle errors/exceptions the same way)</strong></summary>

```diff
diff --git a/src/Symfony/Component/Console/Application.php b/src/Symfony/Component/Console/Application.php
index fd12aa0..6175332 100644
--- a/src/Symfony/Component/Console/Application.php
+++ b/src/Symfony/Component/Console/Application.php
@@ -36,6 +36,7 @@ use Symfony\Component\Console\Event\ConsoleExceptionEvent;
 use Symfony\Component\Console\Event\ConsoleTerminateEvent;
 use Symfony\Component\Console\Exception\CommandNotFoundException;
 use Symfony\Component\Console\Exception\LogicException;
+use Symfony\Component\Debug\ErrorHandler;
 use Symfony\Component\Debug\Exception\FatalThrowableError;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 
@@ -62,7 +63,6 @@ class Application
     private $name;
     private $version;
     private $catchExceptions = true;
-    private $catchErrors = false;
     private $autoExit = true;
     private $definition;
     private $helperSet;
@@ -124,9 +124,10 @@ class Application
                 throw $e;
             }
         } catch (\Error $e) {
-            if (!$this->catchErrors) {
+            if (!$this->catchExceptions) {
                 throw $e;
             }
+
             $e = new FatalThrowableError($e);
         }
 
@@ -278,22 +279,6 @@ class Application
     }
 
     /**
-     * @return bool Whether errors are caught or not during commands execution
-     */
-    public function areErrorsCaught()
-    {
-        return $this->catchErrors;
-    }
-
-    /**
-     * @param bool $boolean Whether to catch errors or not during commands execution
-     */
-    public function setCatchErrors($boolean)
-    {
-        $this->catchErrors = (bool) $boolean;
-    }
-
-    /**
      * Gets whether to automatically exit after a command execution or not.
      *
      * @return bool Whether to automatically exit after a command execution or not
```

</details>


the output changes from:

```sh
[Symfony\Component\Debug\Exception\UndefinedMethodException]
Attempted to call an undefined method named "Option" of class "Symfony\Component\Console\Input\ArgvInput".
Did you mean to call e.g. "getOption", "getOptions", "getParameterOption", "hasOption", "hasParameterOption" or "setOption"?
```

to

```sh
[Error]
Call to undefined method Symfony\Component\Console\Input\ArgvInput::Option() 
```

(Some github service outages is preventing me to upload screens)

If I try to detect the Symfony `ErrorHandler` was registered using this patch:

<details>
<summary><strong>PATCH 2 (detect error handler)</strong></summary>

```diff
diff --git a/src/Symfony/Component/Console/Application.php b/src/Symfony/Component/Console/Application.php
index fd12aa0..ddc81f1 100644
--- a/src/Symfony/Component/Console/Application.php
+++ b/src/Symfony/Component/Console/Application.php
@@ -36,6 +36,7 @@ use Symfony\Component\Console\Event\ConsoleExceptionEvent;
 use Symfony\Component\Console\Event\ConsoleTerminateEvent;
 use Symfony\Component\Console\Exception\CommandNotFoundException;
 use Symfony\Component\Console\Exception\LogicException;
+use Symfony\Component\Debug\ErrorHandler;
 use Symfony\Component\Debug\Exception\FatalThrowableError;
 use Symfony\Component\EventDispatcher\EventDispatcherInterface;
 
@@ -62,7 +63,6 @@ class Application
     private $name;
     private $version;
     private $catchExceptions = true;
-    private $catchErrors = false;
     private $autoExit = true;
     private $definition;
     private $helperSet;
@@ -124,9 +124,18 @@ class Application
                 throw $e;
             }
         } catch (\Error $e) {
-            if (!$this->catchErrors) {
+            if (!$this->catchExceptions) {
+                throw $e;
+            }
+
+            $handler = set_exception_handler('var_dump');
+            $handler = is_array($handler) ? $handler[0] : null;
+            restore_exception_handler();
+            // Symfony Debug error handler was registered, let it handle the error:
+            if ($handler instanceof ErrorHandler) {
                 throw $e;
             }
+
             $e = new FatalThrowableError($e);
         }
 
@@ -278,22 +287,6 @@ class Application
     }
 
     /**
-     * @return bool Whether errors are caught or not during commands execution
-     */
-    public function areErrorsCaught()
-    {
-        return $this->catchErrors;
-    }
-
-    /**
-     * @param bool $boolean Whether to catch errors or not during commands execution
-     */
-    public function setCatchErrors($boolean)
-    {
-        $this->catchErrors = (bool) $boolean;
-    }
-
-    /**
      * Gets whether to automatically exit after a command execution or not.
      *
      * @return bool Whether to automatically exit after a command execution or not

```

</details>

problem is solved using the fullstack. But that's wrong, as it only tells the `ErrorHandler` was set, like in any app using `Debug::enable()`. It doesn't ensure the following has been registered as exception handler (this is what is done by the `DebugHandlersListener`):

```php
function ($e) use ($app, $output) {
    $app->renderException($e, $output);
};
```

So any standalone console app using the Debug component would get the raw output instead of being nicely handled by the console component.

Maybe we could try looking if the `DebugHandlersListener` is registered, but doing such a thing in the `Application` class looks inappropriate.

Also we should remember that despite the issue I'm trying to describe is about Symfony's `ErrorHandler`, someone using its own way to output errors to console would encounter the same issues if we're suddenly handling errors in the `Application` class.

So, yes, I can revert changes to handle errors and exceptions exactly the same way, without this flag (PATCH 1). I just want to be sure everyone is aware of this. To me it's a stumbling block. 😕 

At least, the new flag is opt-in and won't cause more troubles than adding 1 line to enable errors catching in a standalone console app.

However, if anyone has another clean and BC solution regarding this, I'll be glad to give it a try. :)

**Comment by fabpot** [comment]
Is it still relevant? We fixed quite a few bugs related to that recently.

**Comment by ogizanagi** [comment]
I need to check, but last time I did, it was still needed to me as trying to explain in my last comment. 😕 

I didn't follow much changes lately. Which commits should have enhanced the situation regarding this?

File: src/Symfony/Component/Console/Application.php [hunk 6]
```diff
@@ -124,7 +123,14 @@ public function run(InputInterface $input = null, OutputInterface $output = null
             if (!$this->catchExceptions) {
                 throw $e;
             }
+        } catch (\Error $e) {
```
Comment (lstrojny) [inline src/Symfony/Component/Console/Application.php]: Shouldn’t this be `\Throwable`?

Comment (ogizanagi) [inline src/Symfony/Component/Console/Application.php]: Using `Throwable` would make the `catchErrors` flag catch both exceptions and errors, where there already is a dedicated flag for exceptions.
So, if we keep both flags (and I still think we need it currently, as I tried to explain in my previous comments, unless we find a better alternative :/ ), it should probably remains `\Error` here.

**Comment by ogizanagi** [comment]
Since #18140 was merged ( ping @wouterj ), this one does not makes sense anymore as is. But as explained, catching errors indistinctly from exceptions could be considered as a BC break, as error handlers are not used anymore. Using the same `catchExceptions` flag for both exceptions & errors prevents us from having a clean path to ensure the BC.

So you can close this as soon as the BC break is acknowledged but considered acceptable to you. But this is also a DX regression in the case of the FrameworkBundle, which used to register a handler & enhance the messages, but doesn't anymore now. :/

**Comment by chalasr** [comment]
Could you confirm #22261 fixes the BC break? I'm not able to add a relevant test for it

**Comment by nicolas-grekas** [comment]
I think we can close this one, I'm going to revisit the issued spotted here as introduced by #18140 separately.
